### PR TITLE
Params side of the design changes around Add methods

### PR DESCRIPTION
**Pull request by AlekseyTs**
The main observable change - the original (unconverted) params arguments are used in the process of populating target collection. 

Spec changes: https://github.com/dotnet/csharplang/pull/8022

Fixes https://github.com/dotnet/roslyn/issues/72098

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3637]: >bindInterpolatedStringHandlerInMemberCall [](http://example.com/codeflow?start=28&length=41)

This used to be ```BindInterpolatedStringHandlerInMemberCall``` at the end of the `Binder_InterpolatedString.cs`. No meaningful changes during the move.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:877]: >BindInterpolatedStringHandlerInMemberCall [](http://example.com/codeflow?start=32&length=41)

Moved as a local function `bindInterpolatedStringHandlerInMemberCall` to `Binder_Expressions.cs`

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3640]: >ArrayBuilder<BoundExpression>? arguments, [](http://example.com/codeflow?start=16&length=41)

Made this parameter nullable.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3670]: >Debug.Assert(arguments is not null); [](http://example.com/codeflow?start=16&length=36)

Added this assert.

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1523]: >paramsArgsBuilder.Add(argumentsBuilder[i]); [](http://example.com/codeflow?start=24&length=43)

Logic for this and the next statement (the `for`) was moved to local function ```CheckAndCoerceArguments.collectParamsArgs```

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3510]: >collectParamsArgs [](http://example.com/codeflow?start=42&length=17)

This is essentially the logic that used to be in a loop at the beginning of ```BindDefaultArgumentsAndParamsCollection```

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1573]: >TypeSymbol collectionType = parameters[paramsIndex].Type; [](http://example.com/codeflow?start=16&length=57)

Code starting from here through ```Debug.Assert(collection.IsParamsArrayOrCollection);``` is extracted to ```CreateParamsCollection``` helper. No chnages to the code other than making  ```parameters[paramsIndex]``` a parameter named  "paramsParameter".

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1744]: >CreateParamsCollection [](http://example.com/codeflow?start=32&length=22)

This is a helper extracted from ```BindDefaultArgumentsAndParamsCollection```. See a comment at the original location

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:1684]: >for (var i = firstParamsArgument + collectionArgsLength - 1; i != firstParamsArgument; i--) [](http://example.com/codeflow?start=20&length=91)

Logic in this loop is moved to a loop at the end of the ```CheckAndCoerceArguments.createParamsCollection``` local function

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3593]: >for (var i = firstParamsArgument + collectionArgs.Length - 1; i != firstParamsArgument; i--) [](http://example.com/codeflow?start=16&length=92)

Logic in this loop used to be in ```BindDefaultArgumentsAndParamsCollection```. See a comment at the original location.

**Comment by AlekseyTs** [comment]
@cston, @RikkiGibson, @333fred Please review

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3507]: >collectParamsArgs [](http://example.com/codeflow?start=42&length=17)

Can this local function be `static`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3522]: Consider extracting a local function for use here and inside the `while` loop.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3542]: >void [](http://example.com/codeflow?start=12&length=4)

`static`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:9819]: Nit: consider moving this declaration closer to the usage, or inline.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:877]: I would have a slight leaning towards keeping this where it is. I won't block on it, but I did put this here intentionally, rather than making it a local function in CoerceArguments, because I wanted to keep all the interpolation logic together.

**Review by 333fred — APPROVED** [review]
Nothing blocking, so I'm signing off.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:877]: > I would have a slight leaning towards keeping this where it is. I won't block on it, but I did put this here intentionally, rather than making it a local function in CoerceArguments, because I wanted to keep all the interpolation logic together.

The primary reason for the move is to constrain access to ```MemberAnalysisResult,.ArgsToParamsOpt``` to as few helpers that anyone can call in the future as possible. I was able make other helper(s) local functions, I wouldn't otherwise. I find that more important than a desire to keep interpolation logic in a separate file.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3507]: > Can this local function be `static`?

Yes, changing.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3522]: > Consider extracting a local function for use here and inside the `while` loop.

Are you referring to ```paramsArgsBuilder.Add(arguments[arg]);```? If so, I do not believe it is worth the trouble. There isn't enough "critical mass" in combined complexity and volume to gain much from sharing. A call would look as complicated as the current code, in my opinion.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3542]: `coerceArgument` cannot be static, this disallows  ```createParamsCollection``` to be static.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MemberAnalysisResult.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MemberAnalysisResult.cs:39]: The comment was copied from below. I have no intent fixing it in this PR. Feel free to fix both comments after the PR is merged.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: After a closer look, ended up refactoring the loop and avoided duplication.

File: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs:654]: >attributeData.ConstructorArguments.Single() [](http://example.com/codeflow?start=50&length=43)

Is this a meaningful change to the public API behavior for `AttributeData.ConstructorArguments`?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests.cs:654]: > Is this a meaningful change to the public API behavior for `AttributeData.ConstructorArguments`?

Yes. In some error conditions we are actually manage to create an array. The behavior is close to the success case now, see the other test I modified in this file

**Pull request merged**
